File: isabelle.eclipse.launch/src/isabelle/eclipse/launch/config/RootDirLaunch.java
package isabelle.eclipse.launch.config;

import isabelle.eclipse.launch.IsabelleLaunchConstants;
import isabelle.eclipse.launch.IsabelleLaunchPlugin;

import java.io.File;

import org.eclipse.core.runtime.CoreException;
import org.eclipse.debug.core.ILaunchConfiguration;


public class RootDirLaunch extends IsabelleLaunch {

	@Override
	protected String getInstallationPath(ILaunchConfiguration configuration) throws CoreException {
		return getVerifyPath(configuration).getPath();
	}

	private File getVerifyPath(ILaunchConfiguration configuration) throws CoreException {
		
		String location = getLocationConfig(configuration);
		
		if (location == null || location.isEmpty()) {
			abort("Isabelle location not specified");
		}
		
		// allow subclasses to adapt the actual path
		location = adaptPath(location);
		
		File file = new File(location);
		if (!file.exists()) {
			abort("Isabelle location does not exist");
		}
		
		// must be a directory
		if (!file.isDirectory()) {
			abort(getNotDirectoryMessage());
		}
		
		return file;
	}
	
	protected String adaptPath(String path) {
		return path;
	}
	
	protected String getNotDirectoryMessage() {
		return "Isabelle installation location is not a directory";
	}
	
	public static String getLocationConfig(ILaunchConfiguration configuration) {
		String location= ""; //$NON-NLS-1$
		try {
			location= configuration.getAttribute(IsabelleLaunchConstants.ATTR_LOCATION, ""); //$NON-NLS-1$
		} catch (CoreException ce) {
			IsabelleLaunchPlugin.getDefault().log("Error reading configuration", ce);
		}
		return location;
	}
	
}

File: isabelle.eclipse.launch/src/isabelle/eclipse/launch/config/RootDirLaunch.scala
package isabelle.eclipse.launch.config

import java.io.File

import scala.util.Either

import org.eclipse.core.runtime.IStatus
import org.eclipse.debug.core.ILaunchConfiguration

import isabelle.eclipse.launch.config.IsabelleLaunch._
import isabelle.eclipse.launch.config.LaunchConfigUtil._


/**
 * Generic Isabelle launch delegate for directory-based Isabelle installation
 * (e.g. Windows or Linux).
 * 
 * @author Andrius Velykis
 */
class RootDirLaunch extends IsabelleLaunch {

  override def installationPath(configuration: ILaunchConfiguration): Either[IStatus, String] =
    installationDir(configuration).right map (_.getPath)


  private def installationDir(configuration: ILaunchConfiguration): Either[IStatus, File] = {

    val locationConfig = configValue(configuration, IsabelleLaunchConstants.ATTR_LOCATION, "")

    if (locationConfig.isEmpty) {
      abort("Isabelle location not specified")
    } else {

      // allow subclasses to adapt the actual path
      val location = adaptInstallationPath(locationConfig)

      val file = new File(location)

      if (!file.exists) {
        abort("Isabelle location does not exist")

      } else if (!file.isDirectory) {
        // must be a directory
        abort(notDirectoryMessage)

      } else {
        // correct location
        result(file)
      }
    }
  }


  protected def adaptInstallationPath(path: String): String = path

  protected def notDirectoryMessage: String = "Isabelle installation location is not a directory"

}
